lib/aliquot/validator.rb

```diff
@@ -166,59 +160,97 @@ class SignedMessageContract < Dry::Validation::Contract
 
-      rule(:cryptogram) do
-        key.failure('is missing') if 'CRYPTOGRAM_3DS'.eql?(values[:authMethod]) &&
-          values[:cryptogram].nil?
+    class ECv1_PaymentMethodDetailsContract < Dry::Validation::Contract
+      json(CommonPaymentMethodDetailsContract.schema) do
+        required(:pan).filled(:str?)
       end
 
-      rule(:cryptogram) do
-        key.failure('cannot be defined') if 'PAN_ONLY'.eql?(values[:authMethod]) &&
-          !values[:cryptogram].nil?
+      rule(:pan).validate(:integer_string?, :pan?)
+    end
+
+    class ECv1_TokenizedPaymentMethodDetailsContract < Dry::Validation::Contract
+      json(CommonPaymentMethodDetailsContract.schema) do
+        required(:dpan).filled(:str?)
+        required(:threedsCryptogram).filled(:str?)
+        required(:eciIndicator).filled(:str?)
+        required(:authMethod).filled(:str?, included_in?: %w[3DS])
       end
 
-      rule(:eciIndicator) do
-        key.failure('cannot be defined') if 'PAN_ONLY'.eql?(values[:authMethod]) &&
-          !values[:eciIndicator].nil?
+      rule(:dpan).validate(:integer_string?, :pan?)
+      rule(:eciIndicator).validate(:eci?)
+    end
+
+    class ECv2_PaymentMethodDetailsContract < Dry::Validation::Contract
+      json(CommonPaymentMethodDetailsContract.schema) do
+        required(:pan).filled(:str?)
+        required(:authMethod).filled(:str?, included_in?: %w[PAN_ONLY])
       end
+
+      rule(:pan).validate(:integer_string?, :pan?)
     end
 
-    PaymentMethodDetailsSchema = PaymentMethodDetailsContract.new
+    class ECv2_TokenizedPaymentMethodDetailsContract < Dry::Validation::Contract
+      json(CommonPaymentMethodDetailsContract.schema) do
+        required(:pan).filled(:str?)
+        required(:cryptogram).filled(:str?)
+        required(:eciIndicator).filled(:str?)
+        required(:authMethod).filled(:str?, included_in?: %w[CRYPTOGRAM_3DS])
+      end
+
+      rule(:pan).validate(:integer_string?, :pan?)
+      rule(:eciIndicator).validate(:eci?)
+    end
+
+    # PaymentMethodDetailsSchema = PaymentMethodDetailsContract.new
 
     # DRY-Validation schema for encryptedMessage component Google Pay token
     class EncryptedMessageContract < Dry::Validation::Contract
       json do
         required(:messageExpiration).filled(:str?)
         required(:messageId).filled(:str?)
         required(:paymentMethod).filled(:str?)
-        required(:paymentMethodDetails).filled(:hash).schema(PaymentMethodDetailsContract.schema)
+        required(:paymentMethodDetails).filled(:hash)
         optional(:gatewayMerchantId).filled(:str?)
       end
       rule(:messageExpiration).validate(:integer_string?)
+
+      rule(:paymentMethodDetails).validate do
+        contract = nil
+        if 'ECv1'.eql?(values[:protocolVersion])
+          if 'TOKENIZED_CARD'.eql?(values[:paymentMethod])
+            contract = ECv1_TokenizedPaymentMethodDetailsContract.new
+          else
+            contract =  ECv1_PaymentMethodDetailsContract.new
```

Comment:
```suggestion
            contract = ECv1_PaymentMethodDetailsContract.new
```

Reply:
Perhaps a construction like

```
result = if some_cond
           calc_something
         else
           calc_something_else
         end
```

is preferable.

https://rubystyle.guide/#indent-conditional-assignment

:shrug:

Reply:
Yeah, it was a bit sloppy written. I prefer to add return for this rather than conditional assignments.

Reply:
Me too. I don't know if `return` will work in a Dry rule thing.